sound/soc/sof/topology.c

```diff
@@ -3005,6 +3048,12 @@ static void sof_complete(struct snd_soc_component *scomp)
 			break;
 		}
 	}
+	/*
+	 * cache initial values of SOF kcontrols by reading DSP value over
+	 * IPC. It may be overwrite by alsa-mixer after booting up
+	 */
+	/* TODO: cache initial values of SOF kcontrols by topology */
```

Comment:
well that confuses me.
@ranj063 is there a way to get those values from topology, and if yes, why do we read from the DSP?

Reply:
@plbossart this was my thinking too. There are 3 open PR's all addressing kcontrol IO issues and they're all related. Let me spend some time over the next couple of days to connect them all

Reply:
@plbossart @bardliao I looked into this and there are a few tings I want to bring up: 
1. There are no default values specified for volume/switch controls in topology
2. byte controls have the option of adding default data but this is optional
3. Even if there are default values specified in topology, the DSP can override these values based on whether these are valid or not. For example, if the defaul value for volume control exceeds the max supported, the DSP will only set the default value to 0DB atm
4. But overall I agree with the idea that kcontrol IO get could always return the cached value. But this does need the step below to cache the value by reading it once from the DSP. 
5. Lastly, I do not understand what this comment means "TODO: cache initial values of SOF kcontrols by topology"?
For the case of process type components, it is already cached (please refer to jaska's PR).

Reply:
@ranj063 The TODO is originally a reminder to me to get the cached data from topology is there is one. But on the second thought, I think we don't need to do here. We can set the topology's default value to the dsp in topology loading and get the value via ipc here. It can prevent the error if the dsp didn't accept the topology's default value.

Reply:
> @plbossart @bardliao I looked into this and there are a few things I want to bring up:
> 
> 1. There are no default values specified for volume/switch controls in topology
> 2. byte controls have the option of adding default data but this is optional
> 3. Even if there are default values specified in topology, the DSP can override these values based on whether these are valid or not. For example, if the defaul value for volume control exceeds the max supported, the DSP will only set the default value to 0DB atm
> 4. But overall I agree with the idea that kcontrol IO get could always return the cached value. But this does need the step below to cache the value by reading it once from the DSP.
> 5. Lastly, I do not understand what this comment means "TODO: cache initial values of SOF kcontrols by topology"?
>    For the case of process type components, it is already cached (please refer to jaska's PR).

@ranj063 I take your answer as agreement but would like a formal approval before I merge. Thanks